src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -121,142 +116,147 @@ async Task<(ImmutableArray<SerializableImportCompletionItem>, StatisticCounter)>
+            using var _2 = PooledDictionary<INamespaceSymbol, string>.GetInstance(out var namespaceNameCache);
+
+            // Get extension method items from source
+            foreach (var (project, syntaxIndex) in indices.SyntaxIndices!)
             {
-                var project = solution.GetProject(projectId);
-                if (project == null || !project.SupportsCompilation)
-                {
-                    continue;
-                }
+                var filter = CreateAggregatedFilter(targetTypeNames, syntaxIndex);
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var assembly = compilation.Assembly;
 
-                // By default, don't trigger index creation except for documents in current project.
-                var loadOnly = !forceIndexCreation && projectId != currentProject.Id;
-                var cacheEntry = await GetCacheEntryAsync(project, loadOnly, cacheService, cancellationToken).ConfigureAwait(false);
+                var internalsVisible = currentAssembly.IsSameAssemblyOrHasFriendAccessTo(assembly);
 
-                if (cacheEntry == null)
-                {
-                    // Don't provide anything if we don't have all the required SyntaxTreeIndex created.
-                    return GetIndicesResult.NoneResult;
-                }
+                var matchingMethodSymbols = GetPotentialMatchingSymbolsFromAssembly(
+                    compilation.Assembly, filter, namespaceFilter, internalsVisible,
+                    counter, cancellationToken);
 
-                syntaxBuilder.Add(cacheEntry.Value);
+                var isSymbolFromCurrentCompilation = project == currentProject;
+                GetExtensionMethodItemsWorker(position, semanticModel, receiverTypeSymbol, matchingMethodSymbols, isSymbolFromCurrentCompilation, builder, namespaceNameCache);
             }
 
-            // Search through all direct PE references.
-            foreach (var peReference in currentProject.MetadataReferences.OfType<PortableExecutableReference>())
+            // Get extension method items from PE
+            foreach (var (peReference, symbolInfo) in indices.SymbolInfos!)
             {
-                var info = await SymbolTreeInfo.GetInfoForMetadataReferenceAsync(
-                    solution, peReference, loadOnly: !forceIndexCreation, cancellationToken).ConfigureAwait(false);
-
-                if (info == null)
+                var filter = CreateAggregatedFilter(targetTypeNames, symbolInfo);
+                if (currentCompilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol assembly)
                 {
-                    // Don't provide anything if we don't have all the required SymbolTreeInfo created.
-                    return GetIndicesResult.NoneResult;
-                }
+                    var internalsVisible = currentAssembly.IsSameAssemblyOrHasFriendAccessTo(assembly);
```

Comment:
nit: but `IsSameAssembly` couldn't be true here right?  becuase that would mean the compilation had a reference to itself?

Reply:
n/m.